### ISSUE TITLE
Feature request: A command that requests a prompt in the mini buffer

### ISSUE THREAD

**Issue by tmalsburg** [karthink/gptel#20]

**Comment by karthink** [comment]
Sure, this is easy to add. But where would the output show up, and how would you continue the conversation?

**Comment by tmalsburg** [comment]
I would use this in other buffers not in the gptel buffer.  For instance, to create an empty org mode table with 10 rows and 5 columns.  The result of the request should then be inserted at point.  So this is not a conversation scenario but for one-off requests.

**Comment by karthink** [comment]
> For instance, to create an empty org mode table with 10 rows and 5 columns. The result of the request should then be inserted at point.

Can't you do this in the current buffer itself? Type in the prompt, select the line (or not depending on #19) and call `gptel-send` -- this will insert the org-table/response at point.

**Comment by tmalsburg** [comment]
> Can't you do this in the current buffer itself? 

I can but then I have to go back to the prompt and delete it. Too much friction.

**Comment by tmalsburg** [comment]
The problem with GPT integration into Emacs workflows is that the API is pretty slow already and any additional slowdowns make using GPT less attractive.

**Comment by akirak** [comment]
It would be nice if there were a function like this:

```emacs-lisp
(defun gptel-on-region (start end header &optional replace)
  )
```

It is analogous to the built-in `call-process-region` function, and the arguments are defined as follows:

- `start` and `end` specifies the bound of the body which should be usually `region-beginning` and `region-end`.
- `header` is a text prepended to the prompt passed to the API. It can be an instruction such as "refactor this code", "describe this function", "translate this text", etc.
- If `replace` is non-nil, the region will be replaced with the API response. Otherwise, the function will return the response, so the user of the function can use it to his/her liking.

With this function, the user can define a custom command easily. The benefit of integrating with this package is that the user can configure API parameters using the transient interface. Transient allows the user to customize commands using `transient-append-suffix`, etc., so it would be possible for the user to add custom commands to `gptel-send-menu` and so on.

**Comment by karthink** [comment]
@tkfv The idea is great but I'm wary of adding a bunch of `gptel-*` commands. Refactoring isn't a distinct action as far as ChatGPT is concerned, it's just a different prompt/system message. So what's needed is a more convenient way to select the message from the transient menu. I'll add the option to replace the selected region with the results to the menu as well.

On that note, do you have a working example of a prompt that makes ChatGPT return _only_ the rewritten code? Every prompt I've tried results in markdown-formatted code + an explanation.

**Comment by akirak** [comment]
> So what's needed is a more convenient way to select the message from the transient menu. 

I would want to have a function suggested in https://github.com/karthink/gptel/issues/20#issuecomment-1489170520. This will let the user define a bunch of custom commands to fit one's own needs.

With this function, it is easy to define the command discussed in this thread:

```elisp
(defun gtpel-send-region (replace)
  (interactive "P")
  (gptel-on-region (region-beginning) (region-end) (completing-read "Select an action: " gtpel-region-actions) replace))
```

**Comment by karthink** [comment]
> With this function, it is easy to define the command discussed in this thread:

gptel's intended behavior is simple:

- Type a prompt in the current buffer,
- with the response below the prompt in the same buffer.

i.e. just like the web UI.

But between this thread and #30, I've had requests for all possibilities:

- Prompt from minibuffer, response in current buffer.
- Prompt in current buffer, response in echo area.
- Prompt in current buffer, response in a new buffer (in a side window).

I can't support all of these (even with the transient menu), so I'll think about adding something like `gptel-send-region`, but more general where you can supply whatever input and destination you want. The async requirement means it won't be as simple as calling a function, you'll have to supply both a caller and a callback.

**Comment by akirak** [comment]
> more general where you can write some elisp to use whatever input and destination you want

It would be nice if there were a slightly lower API than `gptel-send` that allows a custom payload and callback! A function like `gptel-on-region` is a nice-to-have, but what I wish is to reuse the API communication part of this package as much as possible. This package provides a nice streamlined experience as well as configurability with the transient interface, and nobody would want to re-implement most of it.

**Comment by karthink** [comment]
> It would be nice if there were a slightly lower API than gptel-send that allows a custom payload and callback!

I'm almost there, I'll finish up the API when I next have time to work on it.

**Comment by karthink** [comment]
@akirak I've added `gptel-request`, an API for defining custom behavior. Please see the [wiki](https://github.com/karthink/gptel/wiki).

Let me know if you have any feedback.

**Comment by akirak** [comment]
@karthink Great! The API looks nice. Thank you for making this great package.

**Comment by karthink** [comment]
@tmalsburg 

Option 1

Added transient flags to choose whatever source and destination you need:

![2023-04-09-042930_918x376_scrot](https://user-images.githubusercontent.com/8607532/230770018-9ce87644-6c17-44af-bd39-8c899303dce1.png)

Option 2

For faster access, you can write your own wrapper around `gptel-request`. Please see the [wiki](https://github.com/karthink/gptel/wiki).

**Comment by karthink** [comment]
I'm closing this now, but if you have suggestions on further tweaks to the design for accepting prompts from the minibuffer please post them below. This transient menu is more of a first draft and the behavior could use more testing.
